actionpack/lib/action_controller/metal/rendering.rb

```diff
@@ -78,6 +78,10 @@ def _set_rendered_content_type(format)
         end
       end
 
+      def _set_vary_header
+        self.headers["Vary"] = "Accept" if request.headers["Accept"].present?
```

Comment:
The `Accept` header is included by every browser, so this will (in effect) apply to every request. 

I think https://github.com/rails/rails/blob/a2bd669ed240bec739074188d61ff125e07fa8d1/actionpack/lib/action_dispatch/http/mime_negotiation.rb#L73-L75 might be a better place to add this, since it's the place where the `Accept` header is read by default.

The check for whether to use the accept header is https://github.com/rails/rails/blob/a2bd669ed240bec739074188d61ff125e07fa8d1/actionpack/lib/action_dispatch/http/mime_negotiation.rb#L160-L167 

There's also a global config `config.action_dispatch.ignore_accept_header`, but it can be overridden per controller, and it's already checked on line 74, so I wouldn't use it for this.

Reply:
I'm afraid that it's not possible to add the header there. Because `MimeNegotiation` module is included in `Request` class and it's purely for fetching mime_type/format. Modifying response headers there doesn't make sense and it's hard to do as well.

Reply:
That does make things awkward. 

Perhaps the `Request` class could gain a `response_varies_based_on` method (defaulting to `[]`) which could be overridden in `MimeNegotiation` to add `Accept` (and could also e.g. be overridden for `Vary: content-encoding` when gzipping responses).

Then the header can be added based on that method.

Reply:
I prefer to expose request's `use_accept_header` and `valid_accept_header` instead of adding a new method to do what can already be done with the existing methods. And gzipping responses are handled in `Static` middleware:

https://github.com/rails/rails/blob/a2bd669ed2/actionpack/lib/action_dispatch/middleware/static.rb#L69

So I think we don't need to worry about that now.

Reply:
Those methods aren't called if `params_readable` is true; you'd have to reproduce that check, as wall as calling `use_accept_header` / `valid_accept_header`. 

Also, anyone making future changes to https://github.com/rails/rails/blob/a2bd669ed240bec739074188d61ff125e07fa8d1/actionpack/lib/action_dispatch/http/mime_negotiation.rb#L63 would have to remember (if they're aware in the first place) to mirror their changes to `metal/rendering.rb` (a comment explaining this coupling would probably be sufficient to make sure it happens, though).